### PR TITLE
I221

### PR DESCRIPTION
**Pull request by 1sp**
All the changes till now(excluding /menus). The system is tested using an admin login.

### CODE REVIEW

**Comment by jboz62** [comment]
@1sp Thanks for the latest pull request, Sudhir.

**Comment by jboz62** [comment]
@KathJohnson Katherine, I'll check this out to my computer and test everything out. If I have time tomorrow I can stop by and we can review it.

**Comment by KathJohnson** [comment]
That would be great, thanks.

**Comment by jboz62** [comment]
Hmm, seem to be having some trouble checking this out to my computer. Wonder if it’s because there’s a pull request. It says it’s checked out and I can see I’m on the i221 branch, but the files didn’t update. I’ll have to play with it a bit to see what going on with that.

**Comment by jboz62** [comment]
Ok, I was able to get a local copy of the changes and ran some tests. This looks like a great start, but I’m having issues with the sub-nav not showing up, not sure if that’s all finished or not. I was able to add menu items easily enough, though (see screenshot). Also, we now have the capability to use hyphens in the web page URL, which is great, so "Our Company" can have a URL of /our-company. 

Sudhir, is the sub-nav feature working on your computer? If so, maybe we can do a screen share tomorrow and you can demonstrate it for Katherine, John and I, if you’re available. Let us know and we can arrange a time.

Thank you.


John
--
John Bozaan
Bozaan Design
413-437-0097
www.bozaandesign.com <http://www.bozaandesign.com/>

**Comment by jboz62** [comment]
Another issue I’m seeing is that we’ve lost the portfolios menu functionality, not sure what the situation is there. I still see the old code for it, but it’s not showing up when you add portfolio items. I have 2 portfolio items created and there is nothing showing in the menu:

**Comment by 1sp** [comment]
Hi John, thanks for quickly checking it out. 

For the sub-nav the work flow is:
1) Select a sort order
2) Select menu type "Sub-menu"
3) Select a parent menu item for the sub-nav item
4) Save the page
5) Sub-nav item appears when viewing the parent page (As per the sort order)

Please tell me which step is failing at your end.

**Comment by jboz62** [comment]
Hi Sudhir,

I’m not sure I follow. Here’s what I did to try and get the “our process” page to go under “our company”:



Is that correct? If so, there is no drop down under the menu item when I hover over it on the top menu or click on it in the side menu.

**Comment by 1sp** [comment]
Ok, problem with portfolio is a bit odd because it appears and works in my setup for both themes
portfolio-dark
portfolio-light

I doubt if you have received all the changes from the pull request. Can I please help you in any way?.

Another option is that you can checkout i221 branch from the code I forked (https://github.com/1sp/cms)

**Comment by 1sp** [comment]
I think I got it. Please see if you have set the portfolio menu position in Admin settings

**Comment by jboz62** [comment]
Ah, I see. I’m previewing this in the light theme side, haven’t checked out the dark theme yet. Take a look at that side, you can disable dark theme in the admin settings.

**Comment by jboz62** [comment]
yes, I do, but when I try to set it I get a database error, “column not found" (on light theme). Do I need to do a migrate on my database to get that new setup for the table?

**Comment by 1sp** [comment]
Oh yes, I did created migrations for the database work. I am sorry I forgot to post the setup instructions with the pull request. Please run 'php artisan migrate' to fix this error.

**Comment by jboz62** [comment]
No problem, I forgot to do that, was just realizing that I probably needed to run that. Thank you.

**Comment by 1sp** [comment]
About the sub-nav, I just verified the  code. Actually sub-nav is visible only on dark theme. I am looking into it now

**Comment by jboz62** [comment]
Yes, perfect, portfolios is working fine now, thank you.

**Comment by 1sp** [comment]
Great! I am looking into the issue with sub-nav on the light theme.

**Comment by jboz62** [comment]
Yes, I do see it is working great on dark theme, excellent!

**Comment by 1sp** [comment]
John, I have a concern re: sub-nav functionality on light theme. The code I had cloned at the very start doesn't seems to have section for displaying sub-nav on the light theme template(which is why it is not there in my changes as well). I cannot even see it on ttp://shconstruction2.corbettresearchgroup.com/

Please confirm if this is the case. I can still add it if you can point me to a live page/old code having the necessary markup.

Thanks!

**Comment by jboz62** [comment]
Sudhir, did you make some kind of change on http://shconstruction2.corbettresearchgroup.com/ <http://shconstruction2.corbettresearchgroup.com/>? The site is not displaying correctly now, it’s all messed up. Any idea what happened there? That site is live, so we have to be careful with that one. Please let me know if anything was changed. Thank you.

The only menu item on the light theme that had a drop down was on the portfolios, you can see it here: http://lightthemetest.corbettresearchgroup.com/ <http://lightthemetest.corbettresearchgroup.com/> Click on the top menu item “portfolios” and you will see the drop down.

**Comment by jboz62** [comment]
Sudhir, for testing please use http://lightthemetest.corbettresearchgroup.com/ <http://lightthemetest.corbettresearchgroup.com/> or http://testdeploy.corbettresearchgroup.com/ <http://testdeploy.corbettresearchgroup.com/>. shconstruction2.corbettresearchgroup.com is a live site, it looks like it was switched to light theme, that should be a dark theme site.

**Comment by 1sp** [comment]
John, my mistake. I did changed the theme to check the sub-nav functionality (assuming it was a test site). I will use the test domains which you have mentioned from now on. Sorry again!

**Comment by jboz62** [comment]
Sudhir, dark theme is very confusing, it’s different from light theme in the setup. There shouldn’t be any sub-nav items under portfolios on dark theme, only on light theme. The sub-nav items should only be for posts, which on S+H is the “Builder's Notebook”. So portfolios is actually “projects” and should just show the projects on a grid layout page, and the posts, which I think is called “builder’s notebook” has the sub-nav items.

On light theme, the portfolios item has a drop down menu on the top menu and a slide down menu on the left menu. It’s very confusing.

**Comment by jboz62** [comment]
No problem, I switched it back, looks fine now. Wasn’t sure what happened :)

**Comment by 1sp** [comment]
Ok, this is what I understood from for your message re: dark theme being confusing.

Light theme:

- Portfolio as a drop down in left and top navigation
- No sub-navigation on left navigation

Dark theme:

- No top nav
- Portfolio as Grid page
- Sub-nav appears in left navigation

**Comment by 1sp** [comment]
Glad that http://shconstruction2.corbettresearchgroup.com/ is working again. Thanks John :)

**Comment by jboz62** [comment]
These menus are all done in Bootstrap and just use the basic Bootstrap functionality. If you look at the HTML you will see what I mean:



You can read more about it here: https://getbootstrap.com/components/#navbar <https://getbootstrap.com/components/#navbar>

**Comment by jboz62** [comment]
See it IS confusing :)

Light theme is correct, except there will be sub-nav items for pages on both menus that function like the portfolios menu item.

Dark theme:

No top nav
Portfolio menu item is a grid page (like on S+H) - no sub-navigation items
Sub-navigation for pages.
sub-navigation items on left for posts when you click “Builder’s Notebook” (not sure if “builder’s notebook is used on all dark theme sites. Katherine?).
We only have one dark theme site so far, so not sure if the next one will use “builder’s notebook” for the post. Whatever it’s called, it will have sub-navigation items below it.

**Comment by 1sp** [comment]
Thanks for that patient explanation John, I think I now understand how sub-nav would look on light theme :)

**Comment by 1sp** [comment]
I will work on it today and add the revision to the pull request. Thanks!

File: app/views/shared/top-nav.blade.php [hunk 1]
```diff
@@ -7,38 +7,31 @@
...
-                <li class="{{$active}} dropdown">
-                    <a class="dropdown-toggle"
-                        data-toggle="dropdown"
-                        href="#">
-                        Portfolios
-                    </a>
+      <?php $count = 1; ?>
+      @foreach($top_left_nav as $top)
+        @if($settings->portfolio_menu_postion == $count)
+        <li class="dropdown">
```
Comment (jboz62) [inline app/views/shared/top-nav.blade.php]: @1sp it looks like you edited this section, this is the drop down for portfolios on the top nav. You can see the Bootstrap classes here that you will need to use for creating the drop downs for the pages on the light theme. Probably a similar layout with the classes.

**Comment by jboz62** [comment]
Ok, great, I need to go to bed. Here’s another example of the bootstrap nav: http://www.w3schools.com/bootstrap/tryit.asp?filename=trybs_navbar_dropdown&stacked=h <http://www.w3schools.com/bootstrap/tryit.asp?filename=trybs_navbar_dropdown&stacked=h> The scripts and stylesheets are all included in the sites, just need to use the classes for the menu setup.

**Comment by 1sp** [comment]
Hi John,  this is bit more tricky than what I originally expected. I would need to rewrite the logic of menu generation again.

Current logic assume that there is max one set of sub-nav items for any given page. Now when we wish to get sub nav shown as bootstrap addon , which is making it tricky

**Comment by 1sp** [comment]
I would spend some time to find a work-around for this

**Comment by jboz62** [comment]
Hi Sudhir,

I’m not following you, do you have time for a screen share. Maybe you are misunderstanding me.
> Current logic assume that there is max one set of sub-nav items for any given page
> 
That is correct, it should only be one set of sub-nav items for any given page. The bootstrap is just CSS classes, that’s all you need to include to get the functionality. The bootstrap js and css files are already linked, you just add the classes and the functionality is already there.



For light theme you would have on the top and left menus, for example:

Our Company
Sub-item 1 - drop down
Sub-item 2 - drop down
Sub-item 3 - drop down
Etc.

For dark theme you would just have the sub-navigation as we discussed:

Our Company
________________________

Sub-item 1
Sub-item 2
Sub-item 3
Etc.

Let me know if that’s not clear and we can do a screen share to discuss further.

Thank you,
John

**Comment by 1sp** [comment]
Yes, let's do a screen share. Are you available now ?

**Comment by jboz62** [comment]
Let’s shoot for 11:45 EST. Are you having issues with the setup?

**Comment by 1sp** [comment]
Ok, My agenda would be to discuss the sub-nav on light theme. The way the logic works right now support only one instance of sub-nav on a url. I would be able to explain better on a screen share

**Comment by jboz62** [comment]
Ok, sounds good. Actually, can we shoot for 12EST? Troubleshooting a problem with a website. Please send me a link at noon for Join.me <http://join.me/> or whatever you’re using.

Thank you.

**Comment by 1sp** [comment]
Great. I will share a join.me link with you before 12EST

**Comment by 1sp** [comment]
Hi John, here is the join.me link - https://join.me/651-087-805

**Comment by jboz62** [comment]
Hi Sudhir,

It looks like you were doing some work on the menu sort over the weekend. What work did you do on this? 

Can you hold off on any further work until I get approval on the new plan? I need to make sure we are on track with the budgeting for this. It looks like you have 8 hours in total, now. How many more did you say this would take?

Thank you,
John

**Comment by 1sp** [comment]
Hi John,

I worked on initial integration and configuration of new component for /menus page (https://github.com/ilikenwf/nestedSortable). I believe we already have an agreement on this.

In our last discussion , I estimated around 14 hours of total remaining effort for all the pending work , which includes:

1. Replace /menus page's flat list sorting interface with a hierarchical menu management interface (6 hours).
2. Separate out the menu core to its own datastore(table) to make it more flexible and to streamline  extraction of full/partial menu tree from database (7 hours).
3. To fix sub-nav rendering on light theme (1 hour after completing no.2)
4. #222 (fixed but blocked by i221 branch's pull request)

I am holding off work till your next instruction. Can you please get a confirmation on how should I proceed next.

Thanks,
Sudhir

**Comment by jboz62** [comment]
Hi Sudhir,

The initial agreement was for a different setup that was going to take much less time. There are budget constraints that need to be addressed before moving ahead on the new plan. I don’t set the budget, so we need approval to move forward with this new setup. My understanding was this was going to be 14 hours total time, including the 6 you worked before (balance 8 hours). Are you saying this will be 14 hours beyond what you’ve already worked on this? We’ll need to get approval on that, that’s a much bigger commitment than we originally agreed to. We may need to forego the drop down feature until later to cut costs. Let’s wait to hear back from Katherine.

Thank you,
John

**Comment by 1sp** [comment]
Thanks John for putting forward your concern and I understand it totally.

For the current pending work items my estimate would be at-least 14 hours . I , though, agree on the fact that initial work scope was much less and lacked no. 1 and 2 - hence the low estimate. 

If budget is a constraint, we can go live the current setup (keeping /menus and menu backend as they are right now). I can quickly put a hack in place to sort of sub-nav on light theme so that it works on both themes. I can do it in an hour.

Happy to answer further questions


Thank you,
Sudhir

**Comment by meetinghouse** [comment]
Thanks to both of you for your work and consideration of the budget.

Is there someplace I can view the current state of things now? If I can see
where we are I can give a more intelligent answer as to where to go from
here.

Are the changes to this point in place on testdeploy and/or lightthemetest?

Sincerely,
Katherine Johnson, Research Director
Corbett Research Group
http://www.restorationtrades.com
"Web Marketing that Works"

Publishers of:

The Custom Building & Restoration Trades Directory
http://www.restorationtradesdirectory.com

The Building Arts Notebook, "Journal of the Guild of Building Artisans"
http://buildingartisansguild.com/

Email:  katherine@restorationtrades.com
Phone: (413) 475-3154

**Comment by 1sp** [comment]
I believe John is in better position to answer this as I am totally unaware of the deployment workflow. All my changes are already part of the pending pull request from me.

**Comment by jboz62** [comment]
John,

The changes haven’t been deployed to any site yet, I have pulled down the changes to my local copy to test out. I think Sudhir has more changes to committed to the repository, so I’d like to get those changes to my computer as well, then I can come by and we can do a walk through to show you the current state.

Sudhir, can you delete the latest pull request and commit any changes you have for us to test? I think instead of doing a pull request, just yet, please commit the changes to a new branch off of “blog” that I can pull down to my computer for testing and to show John and Katherine.

Thank you,
John B.

**Comment by jboz62** [comment]
As per my previous email, please delete the pull request i221 (can you leave the branch?) and commit any additional changes you have for review in a *new branch* (no pull request). If that is approved, we’ll let you know and you can do a pull request.

Thank you,
John

**Comment by jboz62** [comment]
Yes, please do the “hack” you recommended and then commit all changes to a new branch that we can test, no need for a pull request at this time.

Thank you,
John

**Comment by 1sp** [comment]
Closing this pull request as requested by John.

**Comment by jboz62** [comment]
Hi Sudhir,

Any progress update on the latest commit changes?

Thank you,
John

**Pull request closed without merging**